finetune/nn/target_blocks.py

```diff
@@ -517,6 +517,185 @@ def seq_lab_internal(hidden):
+
+                if use_crf:
+                    log_likelihood, _ = crf_log_likelihood(
+                        logits, targets, lengths, transition_params=transition_params
+                    )
+                    loss = -log_likelihood
+                else:
+                    weights = tf.sequence_mask(
+                        lengths, maxlen=tf.shape(input=targets)[1], dtype=tf.float32
+                    ) / tf.expand_dims(tf.cast(lengths, tf.float32), -1)
+                    loss = tf.compat.v1.losses.sparse_softmax_cross_entropy(
+                        targets,
+                        logits,
+                        weights=weights
+                    )
+
+                    e = 0.02
+                    k = 1
+                    probs = tf.stop_gradient(tf.nn.softmax(all_logits))
+                    adv_vector = tf.random.uniform(shape=tf.shape(preturb_target),
+                                                   dtype=tf.float32)
+                    adv_vector = e * tf.nn.l2_normalize(adv_vector, axis=-1)
+                    batch_size = tf.cast(tf.shape(preturb_target)[0], tf.float32)
+                    kl_div = tf.losses.KLDivergence(reduction=tf.keras.losses.Reduction.NONE)
+                    mask = tf.sequence_mask(all_lengths,
+                                            maxlen=tf.math.reduce_max(all_lengths),
+                                            dtype=tf.float32)
+                    for _ in range(k):
+                        preturbed_hidden = preturb_target + adv_vector
+                        adv_logits = out_fn(preturbed_hidden)
+                        adv_probs = tf.nn.softmax(adv_logits)
+                        adv_loss = kl_div(probs, adv_probs, sample_weight=mask)
+                        adv_loss = tf.reduce_sum(adv_loss) / batch_size
+                        gradient = tf.gradients(adv_loss, adv_vector)
+                        adv_vector = e * tf.nn.l2_normalize(gradient, axis=-1)
+                        adv_vector = tf.reshape(adv_vector,
+                                                tf.shape(adv_vector)[1:])
+                        adv_vector = tf.stop_gradient(adv_vector)
+                        # adv_vector = tf.compat.v1.Print(adv_vector,
+                        #                                 [tf.shape(hidden),
+                        #                                  tf.shape(preturbed_hidden),
+                        #                                  tf.shape(adv_logits),
+                        #                                  tf.shape(adv_probs),
+                        #                                  tf.shape(adv_loss),
+                        #                                  tf.shape(gradient),
+                        #                                  tf.shape(adv_vector)],
+                        #                                summarize=100)
+
+                    preturbed_hidden = preturb_target + adv_vector
```

Comment:
check the variances of this resulting vector to check that adv_vector is noise and not the dominant signal.